### ISSUE TITLE
Add printf folding when the arguments are constant

### ISSUE THREAD

**Issue by michaelrj-google** [llvm/llvm-project#85450]
Existing printf optimizations are very narrow, only handling format strings with no conversions or a single, narrowly defined conversion (%c or %s). This leads to missed optimizations when the printf formatting could be done at compile time, such as for integer conversions or multiple strings.

example: https://godbolt.org/z/e498c1Mhx

My suggestion for how to solve this problem is to check if the printf format string is valid and matches the arguments (similar to what is done for the `-Wformat` warning), then check if the arguments are known at compile time. If both conditions are met, then it should be possible to just run sprintf at compile time and replace the printf call in the generated code with a puts (or fputs) call on a constant string.

cc: @nickdesaulniers @efriedma-quic

**Comment by Poseydon42** [comment]
I would like to work on this issue.

**Comment by nickdesaulniers** [comment]
@Poseydon42 you'll want to take a look at `LibCallSimplifier::optimizePrintF` (https://github.com/llvm/llvm-project/blob/66125ad8e9978e6fdec0e1c3ff83e717eff07aca/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L3117)

**Comment by Poseydon42** [comment]
Yes, I've started looking into that function, and I think I have a general idea of how to approach this. I was wondering if there's an already existing functionality for validating `printf` format strings, or if I need to reimplement it myself. I believe I can simply use `sprintf` to actually fold `printf` when all arguments are constant, but the problem becomes a bit more difficult when only some of the arguments are constant - e.g. `printf("%s %s", "Hello, ", username)`. I believe that the later problem can be solved as well, but I'm not sure if it's not going to cause issues with the expected return value of `printf` and, again, whether this or similar funcionality already exists somewhere, or if I need to write it myself.
